src/psyclone/psyad/main.py

```diff
@@ -0,0 +1,143 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2021, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''Top-level driver functions for PSyAD : the PSyclone Adjoint
+support. Transforms an LFRic tangent linear kernel to its adjoint.
+
+'''
+from __future__ import absolute_import
+import argparse
+import logging
+from io import StringIO
+import sys
+
+import six
+
+from psyclone.psyir.frontend.fortran import FortranReader
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.generator import write_unicode_file
+from psyclone.psyad import generate
+
+
+def main(args):
+    '''Takes an LFRic tangent linear kernel source file as input and
+    produces its adjoint.
+
+    :param list args: the list of command-line arguments that PSyAD has \
+                      been invoked with.
+
+    '''
+    parser = argparse.ArgumentParser(
+        description="Run the PSyclone adjoint code generator on an LFRic "
+        "tangent-linear kernel file")
+    parser.add_argument(
+        '-v', '--verbose', help='increase the verbosity of the output',
+        action='store_true')
+    parser.add_argument('-oad', help='filename for the transformed code')
```

Comment:
I know this is consistent with how vanilla PSyclone does things but would it make more sense just to support, e.g.:

    psyad my_tl_kernel_mod.f90 my_adj_tl_kernel_mod.f90

i.e. just have the input and output filenames as mandatory arguments.

Reply:
I'm not convinced. I think it might be useful for the user to be able to see the output quickly if they are developing/testing, rather than having to look in an output file. How about we ask the users what they think?

Reply:
I didn't say, but I meant that if no output file was provided, then the script would still output to stdout. However, asking the users sounds like a good plan.

Reply:
TBH, having just read the new documentation (thanks) I don't think it's worth holding this PR up for this.